include/kokkos-utils/callbacks/TimerListener.hpp

```diff
@@ -0,0 +1,209 @@
+
+    //! When the listener has matched both the begin and end events, it is said to be "closed".
+    bool closed() const { return this->is_closed; }
+
+    //! Open the listener so that it can be reused to match a begin event again.
+    void open()
+    {
+        this->event_id = invalid_event_id;
+        this->is_closed = false;
+    }
+
+public:
+    Kokkos::utils::timer::Timer<Exec> timer {};
+
+private:
+    Exec exec;
+    uint32_t dev_id;
+    MatcherType matcher;
+    uint64_t event_id = invalid_event_id;
+    bool is_closed = false;
+};
+
+/**
+ * @brief Listener that starts a timer when a @ref Kokkos::utils::callbacks::PushRegionEvent is received
+ *        and stops it when a corresponding @ref Kokkos::utils::callbacks::PopRegionEvent is received.
+ *
+ * The start event is selected by requiring it to be of type @ref Kokkos::utils::callbacks::PushRegionEvent
+ * and to be a match for @ref matcher. After the start event is received, the listener counts subsequent
+ * @ref Kokkos::utils::callbacks::PushRegionEvent events, corresponding to nested timing regions, and it
+ * decreases this counter when a pop event is received. The timer is stopped when this
+ * counter becomes zero.
+ *
+ * This class is well-suited for timing the execution of a region of code enclosed by
+ * a pair of @c Kokkos::Profiling::pushRegion and @c Kokkos::Profiling::popRegion calls,
+ * or enclosed in a @c Kokkos::Profiling::ScopedRegion.
+ *
+ * @note The use of a counter to track nested regions and stop the timer when the counter
+ *       reaches zero is necessary because @c Kokkos does not provide a way to match
+ *       push and pop region events directly.
+ */
+template <
+    Kokkos::utils::callbacks::Matcher<Kokkos::Impl::type_list<Kokkos::utils::callbacks::PushRegionEvent>> MatcherType,
+    Kokkos::utils::concepts::ExecutionSpace Exec
+>
+class TimerListener<Kokkos::utils::callbacks::PushRegionEvent, Kokkos::utils::callbacks::PopRegionEvent, MatcherType, Exec>
+{
+public:
+    static constexpr uint32_t invalid_nested_level = Kokkos::Experimental::finite_max_v<uint32_t>;
+
+public:
+    template <typename T> requires (std::same_as<std::remove_cvref_t<T>, MatcherType>)
+    explicit TimerListener(T&& matcher_)
+      : matcher(std::forward<T>(matcher_)) {}
+
+    template <typename T> requires (std::same_as<std::remove_cvref_t<T>, MatcherType>)
+    explicit TimerListener(Exec exec_, T&& matcher_)
+      : exec(std::move(exec_)), matcher(std::forward<T>(matcher_)) {}
+
+    void operator()(const Kokkos::utils::callbacks::PushRegionEvent& event)
+    {
+        if (matcher(event)) {
+            if (this->connected() || this->closed()) Kokkos::abort("Listener cannot match a region twice.");
+
+            this->timer.reset(this->exec);
+            this->nested_level = 1;
+        } else {
+            if (this->connected()) ++this->nested_level;
+        }
+    }
+
+    void operator()(const Kokkos::utils::callbacks::PopRegionEvent&)
```

Comment:
```suggestion
    void operator()(const PopRegionEvent&)
```

and similar occurrences in this file.

Reply:
Outdated.